In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
import nltk
nltk.download('vader_lexicon')

[nltk_data] Downloading package vader_lexicon to
[nltk_data]     C:\Users\91999\AppData\Roaming\nltk_data...


True

In [3]:
from nltk.sentiment.vader import SentimentIntensityAnalyzer

In [5]:
sid=SentimentIntensityAnalyzer()

In [6]:
df = pd.read_csv('../TextFiles/amazonreviews.tsv', sep='\t')
df.head()

label                                             review
0   pos  Stuning even for the non-gamer: This sound tra...
1   pos  The best soundtrack ever to anything.: I'm rea...
2   pos  Amazing!: This soundtrack is my favorite music...
3   pos  Excellent Soundtrack: I truly like this soundt...
4   pos  Remember, Pull Your Jaw Off The Floor After He...

In [7]:
df['label'].value_counts()

neg    5097
pos    4903
Name: label, dtype: int64

In [10]:
#Lets remove nan values or blank records
df.dropna(inplace=True)

blanks=[]
for i,lb,rv in df.itertuples():
    if type(rv)==str:
        if rv.isspace():
            blanks.append(i)
                
df.drop(blanks,inplace=True)

In [11]:
df['label'].value_counts()

neg    5097
pos    4903
Name: label, dtype: int64

In this case there are no empty reords

In [13]:
sid.polarity_scores(df.iloc[0].review)

{'neg': 0.088, 'neu': 0.669, 'pos': 0.243, 'compound': 0.9454}

we'll add columns to the original DataFrame to store polarity_score dictionaries, extracted compound scores, and new "pos/neg" labels derived from the compound score. We'll use this last column to perform an accuracy test.

In [14]:
df['scores']=df['review'].apply(lambda r:sid.polarity_scores(r))

In [15]:
df.head()

label                                             review  \
0   pos  Stuning even for the non-gamer: This sound tra...   
1   pos  The best soundtrack ever to anything.: I'm rea...   
2   pos  Amazing!: This soundtrack is my favorite music...   
3   pos  Excellent Soundtrack: I truly like this soundt...   
4   pos  Remember, Pull Your Jaw Off The Floor After He...   

                                              scores  
0  {'neg': 0.088, 'neu': 0.669, 'pos': 0.243, 'co...  
1  {'neg': 0.018, 'neu': 0.837, 'pos': 0.145, 'co...  
2  {'neg': 0.04, 'neu': 0.692, 'pos': 0.268, 'com...  
3  {'neg': 0.09, 'neu': 0.615, 'pos': 0.295, 'com...  
4  {'neg': 0.0, 'neu': 0.746, 'pos': 0.254, 'comp...

In [17]:
df['scores'][1]

{'neg': 0.018, 'neu': 0.837, 'pos': 0.145, 'compound': 0.8957}

In [19]:
df['compund']=df['scores'].apply(lambda score_dict:score_dict['compound'])

In [20]:
df.head()

label                                             review  \
0   pos  Stuning even for the non-gamer: This sound tra...   
1   pos  The best soundtrack ever to anything.: I'm rea...   
2   pos  Amazing!: This soundtrack is my favorite music...   
3   pos  Excellent Soundtrack: I truly like this soundt...   
4   pos  Remember, Pull Your Jaw Off The Floor After He...   

                                              scores  compund  
0  {'neg': 0.088, 'neu': 0.669, 'pos': 0.243, 'co...   0.9454  
1  {'neg': 0.018, 'neu': 0.837, 'pos': 0.145, 'co...   0.8957  
2  {'neg': 0.04, 'neu': 0.692, 'pos': 0.268, 'com...   0.9858  
3  {'neg': 0.09, 'neu': 0.615, 'pos': 0.295, 'com...   0.9814  
4  {'neg': 0.0, 'neu': 0.746, 'pos': 0.254, 'comp...   0.9781

In [23]:
df['predlabel']=df['compund'].apply(lambda s:'pos' if s>0 else 'neg')

In [24]:
df.head()

label                                             review  \
0   pos  Stuning even for the non-gamer: This sound tra...   
1   pos  The best soundtrack ever to anything.: I'm rea...   
2   pos  Amazing!: This soundtrack is my favorite music...   
3   pos  Excellent Soundtrack: I truly like this soundt...   
4   pos  Remember, Pull Your Jaw Off The Floor After He...   

                                              scores  compund predlabel  
0  {'neg': 0.088, 'neu': 0.669, 'pos': 0.243, 'co...   0.9454       pos  
1  {'neg': 0.018, 'neu': 0.837, 'pos': 0.145, 'co...   0.8957       pos  
2  {'neg': 0.04, 'neu': 0.692, 'pos': 0.268, 'com...   0.9858       pos  
3  {'neg': 0.09, 'neu': 0.615, 'pos': 0.295, 'com...   0.9814       pos  
4  {'neg': 0.0, 'neu': 0.746, 'pos': 0.254, 'comp...   0.9781       pos

In [25]:
from sklearn.metrics import accuracy_score,classification_report,confusion_matrix

In [26]:
accuracy_score(df['label'],df['predlabel'])

0.7122

In [28]:
print(classification_report(df['label'],df['predlabel']))

              precision    recall  f1-score   support

         neg       0.85      0.53      0.65      5097
         pos       0.65      0.90      0.75      4903

    accuracy                           0.71     10000
   macro avg       0.75      0.72      0.70     10000
weighted avg       0.75      0.71      0.70     10000



Vader roughly identifies the sentiment correctly 71% of time